### DISCUSSION TITLE
Questions about camera resolutions and how to get the old "modes" of the old picamera-lib?

### DISCUSSION THREAD

**Discussion by cpixip** [raspberrypi/picamera2#157] [Q&A] [answered]
While having fun in understanding the new picamera2-lib and trying to adapt the new lib to my existing software suite (based on the old picamera-lib) I stumbled upon a few things I could not figure out by reading/searching the examples and documentation.

(System: RP4 with HQ sensor, socket-based client-server structure, MJPEG-encoded 2016x1520 frames (q=90) send from the RP4 to a WIN10 machine at 23 fps. Bandwidth used about 130 MBit/s)

First, requesting an image resolution like this,

`video_config = picam2.video_configuration(main={"size": (2028,1520)})`

it seems I only get 2016 x 1520 px in the end. Libcamera reports:

`Selected sensor format: 2028x1520-SGRBG12_1X12 - Selected unicam format: 2028x1520-pgCC`

So I think the camera is using the 2x2 binned mode here, but is cropping the image somewhat.

The same thing kind of happens if I request the full sensor resolution:

`still_config = picam2.still_configuration(main={"size": picam2.sensor_resolution})`

and libcamera again reports

` Selected sensor format: 4056x3040-SGRBG12_1X12 - Selected unicam format: 4056x3040-pgCC`

as requested. 

But: I only get 4032 x 3040 px images. (I needed here to use the still_configuration because the video_configuration ran out of buffer space). So again, the delivered image size is slightly smaller in x-dimension than the requested one.

Has this behaviour something to do with buffer alignments or so? Is it caused by libcamera or the encoder used (simplejpeg)?

A further question: the old picamera-lib (based on the Broadcom-stack) allowed to run the sensor at full resolution (`mode=3`) and half the output image size to 2028 x 1520 by rescaling. This yielded a noticably better image quality than running the HQ sensor in `mode=2` with it's 2x2 binning.  Here's an example of the difference (obtained with the old picamera/Broadcom stuff):

![grafik](https://user-images.githubusercontent.com/43588739/171999768-2481669c-7b4b-4bee-be0d-9e3b7631152e.png)

How can I request from libcamera the use of the non-binned 4056 x 3040 sensor mode while working with an output image size of 2028 x 1520 or so?

**Comment by sandyol55** [comment, marked as answer]
Think the recommended method is to specify an additional  _raw_ configuration at the full sensor resolution along with the _main_ 1/2 resolution required image size.

Not tested on an HQ camera, but with a V2.1 the following configuration drives the sensor into its max resolution mode but outputs a rescaled 'full format but  1/2 pixel density per axis'  image.

`still_config = picam2.still_configuration(main={"size" : (1640,1232}), raw={"size" : (3280,2464)})`

Interestingly the achieved image size is also reduced horizontally, by 8 pixels in this case,  for a 1632 x 1232 image

**Reply by cpixip** [reply]
thank you! That does the trick. Seems that the horizontal size of a delivered image is always quantized to 16 in our case. There is the `align_stream(self, stream_config: dict)`-fct in the code which does such a thing with certain image formats, but I am explicitly requesting "RGB888" which is not in this list. Anyway, thanks for resolving my "mode"-question!

**Reply by sandyol55** [reply]
Glad it worked, the tip came from an answer I'm sure DP has given elsewhere but I couldn't remember where, to link to it.
Seeing your USAF three bar charts brought back some early memories of QA testing 35mm projection systems for a moving map system!

**Reply by cpixip** [reply]
:+1:

**Reply by davidplowman** [reply]
A couple of points that might interest folks here:

* You can find out what raw camera modes are available using `libcamera-hello --list-cameras`. To have that available in Picamera2 is a feature on the list. I'd like to provide a bit more information about the modes, but whether that's possible is still TBD.
* When you ask for a camera configuration, it gives you a suggested configuration, but you can still change it. For example, it forces widths to be aligned to how the hardware works so that you get less copying of multi-megapixel buffers. The assumption is that the optimal alignment rules are a bit tricky, so it's best if it gets done for you. But you can ask for other sizes if you want: just do `config['main']['size'] = (4056, 3040)` once the suggested configuration has been returned to you.

I hope a lot of this will become clearer once there's official documentation (which I'm just starting on).

**Reply by davidplowman** [reply]
Just to add to this that the latest versions of _Picamera2_ have a `sensor_modes` property, which you can use like this:
```
from picamera2 import Picamera2
from pprint import *
picam2 = Picamera2()
pprint(picam2.sensor_modes)
```
It's best to query the `sensor_modes` property immediately after creating the `Picamera2` object because it will try to reconfigure the camera multiple times, but after that it is cached. You can even write things like
```
config = picam2.create_preview_configuration(raw=picam2.sensor_modes[0])
picam2.configure(config)
```
